servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Single.java

```diff
@@ -1863,6 +1888,118 @@ public final Cancellable subscribe(Consumer<? super T> resultConsumer) {
+     * <pre>{@code
+     *      CompletableFuture<T1> f1 = ...; // s1
+     *      CompletableFuture<T2> f2 = ...; // s2
+     *      CompletableFuture<T3> f3 = ...; // s3
+     *      CompletableFuture.allOf(f1, f2, f3).get(); // wait for all futures to complete
+     *      return zipper.apply(f1.get(), f2.get(), f3.get());
+     * }</pre>
+     * @param s1 The first {@link Single} to zip.
+     * @param s2 The second {@link Single} to zip.
+     * @param s3 The third {@link Single} to zip.
+     * @param zipper Used to combine the completed results for each item from {@code singles}.
+     * @param <T1> The type for the first {@link Single}.
+     * @param <T2> The type for the second {@link Single}.
+     * @param <T3> The type for the third {@link Single}.
+     * @param <R> The result type of the combinator.
+     * @return a new {@link Single} that emits the results of a specified combinator {@link Function} to items emitted
+     * by {@code singles}.
+     * @see <a href="http://reactivex.io/documentation/operators/zip.html">ReactiveX zip operator.</a>
+     */
+    public static <T1, T2, T3, R> Single<R> zip(
+            Single<? extends T1> s1, Single<? extends T2> s2, Single<? extends T3> s3,
+            Function3<? super T1, ? super T2, ? super T3, ? extends R> zipper) {
+        return SingleZipper.zip(s1, s2, s3, zipper);
+    }
+
+    /**
+     * Create a new {@link Single} that emits the results of a specified combinator {@link Function} to items emitted
+     * by {@code singles}.
+     * <p>
+     * From a sequential programming point of view this method is roughly equivalent to the following:
+     * <pre>{@code
+     *      CompletableFuture<T1> f1 = ...; // s1
+     *      CompletableFuture<T2> f2 = ...; // s2
+     *      CompletableFuture<T3> f3 = ...; // s3
+     *      CompletableFuture<T4> f4 = ...; // s3
+     *      CompletableFuture.allOf(f1, f2, f3, f4).get(); // wait for all futures to complete
+     *      return zipper.apply(f1.get(), f2.get(), f3.get(), f4.get());
+     * }</pre>
+     * @param s1 The first {@link Single} to zip.
+     * @param s2 The second {@link Single} to zip.
+     * @param s3 The third {@link Single} to zip.
+     * @param s4 The forth {@link Single} to zip.
+     * @param zipper Used to combine the completed results for each item from {@code singles}.
+     * @param <T1> The type for the first {@link Single}.
+     * @param <T2> The type for the second {@link Single}.
+     * @param <T3> The type for the third {@link Single}.
+     * @param <T4> The type for the forth {@link Single}.
+     * @param <R> The result type of the combinator.
+     * @return a new {@link Single} that emits the results of a specified combinator {@link Function} to items emitted
+     * by {@code singles}.
+     * @see <a href="http://reactivex.io/documentation/operators/zip.html">ReactiveX zip operator.</a>
+     */
+    public static <T1, T2, T3, T4, R> Single<R> zip(
```

Comment:
Do we think the 3 and 4 cases are going to be common enough to warrant not jumping immediately from 2 -> many? I usually think of these things in terms of method references. What existing functions with the correct arity are likely to be reused as the zipper? If no functions exist, would the new functions be written only for use a zippers or otherwise be useful. 
Some libraries go to even higher arity than four to get the compile type safety.  The typical var args implementation would be purpose built to size check the input array and then read and cast args invoking some method to produce the R type.

Reply:
> Do we think the 3 and 4 cases are going to be common enough

Usage is difficult to predict but other rx frameworks do have 8+ arity overrides for this method. The fall-back `n`-arity lack-of generics is pretty verbose/cast-heavy in this case, imho I think it is worth a few extra interfaces to reduce the scenarios where the fall-back operator is necessary.

> would the new functions be written only for use a zippers or otherwise be useful.

rxjava has similar combinator and it is used for `zip`, `combine*`, `withLatestFrom`. I assume `zip` is the most popular out of those.